Offloaded KV Cache

n17s opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #30704 

This PR introduces OffloadedCache. This is a KV cache implementation that reduces GPU memory usage in exchange for more CPU memory usage and a small increase in generation time. During the forward passes in `generate`, it only keeps two layers of KV cache on the device: the current layer and the next layer. All other layers are on the CPU and are prefetched/evicted as necessary.

It can be used by passing `cache_implementation="offloaded"` in the `GenerationConfig` like this:
```python
        gen_config = GenerationConfig(
            cache_implementation="offloaded", 
            # other generation options such as 
            num_beams=4,
            num_beam_groups=2,
            num_return_sequences=4,
            diversity_penalty=1.0,
            max_new_tokens=50,
            early_stopping=True,
        )

        outputs = model.generate(
            inputs["input_ids"],
            generation_config=gen_config
        )
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Issue #30704
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR. @ArthurZucker @gante 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

n17s committed:
Initial implementation of OffloadedCache

n17s committed:
enable usage via cache_implementation

gante commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,123 @@ def batch_select_indices(self, indices: torch.Tensor):
+
+        return self.key_cache[layer_idx], self.value_cache[layer_idx]
+
+    @classmethod
+    def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None) -> "OffloadedCache":
+        """Converts a cache in the legacy cache format into an equivalent `DynamicCache`."""
+        cache = cls()
+        if past_key_values is not None:
+            for layer_idx in range(len(past_key_values)):
+                key_states, value_states = past_key_values[layer_idx]
+                cache.key_cache.append(key_states)
+                cache.value_cache.append(value_states)
+                cache.original_device.append(key_states.device)
+        return cache
+
+    def to_legacy_cache(self) -> Tuple[Tuple[torch.Tensor], Tuple[torch.Tensor]]:
+        """Converts the `OffloadedCache` instance into the its equivalent in the legacy cache format."""
+        legacy_cache_lst = []
+        for layer_idx in range(len(self)):
+            original_device = self.original_device[layer_idx]
+            self.key_cache[layer_idx] = self.key_cache[layer_idx].to(original_device)
+            legacy_cache_lst.append((self.key_cache[layer_idx], self.value_cache[layer_idx]))
+        return tuple(legacy_cache_lst)
```
I would overload these two functions with a `NotImplementedError` exception -- they are exclusively used for backwards compatibility, and we would like to get rid of them in the future 🤗

n17s replied:
I have removed these implementations but did it the way it is recommended in the Python docs and left a comment in the code with a pointer to the docs.

gante reviewed (commented):
@n17s The core of the PR LGTM, thank you for proposing and implementing this cool feature 💛 

To finalize this PR:
1. Let's add an integration test (see [this file](https://github.com/huggingface/transformers/blob/main/tests/test_cache_utils.py) for examples). In particular, I would test that a) using this cache yields the same results as the original `DynamicCache`; b) if we're on a GPU, we should see a lower memory consumption than with `DynamicCache`
2. If you have benchmarks (and a benchmark script), please share it in this PR for future reference!

n17s committed:
Address feedback, add tests, remove legacy methods.

n17s commented:
I have incorporated the comment about removing the two legacy implementations and added a couple of tests that ensure the cache works the same way as the DynamicCache and its peak memory usage is lower.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31325). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

n17s commented:
Is something holding this back? The failing tests are all marked as flaky.

gante commented on tests/test_cache_utils.py:
```diff
@@ -455,3 +456,58 @@ def test_static_cache_extra_left_padding(self):
+        model_name = "microsoft/Phi-3-mini-4k-instruct"
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name, device_map="auto", torch_dtype=torch.float16, attn_implementation="flash_attention_2"
```
```suggestion
            model_name, device_map="auto", torch_dtype=torch.float16
```
FA2 places additional requirements on the test. Given that it is not relevant for the test itself, we should remove it 🤗

n17s replied:
I have removed flash attention from the tests.

gante reviewed (approved):
LGTM, thank you for iterating 🤗

n17s committed:
Remove flash-attn, discover synchronization bugs, fix bugs

n17s commented on tests/test_cache_utils.py:
```diff
@@ -455,3 +456,58 @@ def test_static_cache_extra_left_padding(self):
+        model_name = "microsoft/Phi-3-mini-4k-instruct"
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name, device_map="auto", torch_dtype=torch.float16, attn_implementation="flash_attention_2"
```
Fixed.

n17s commented:
> just the limited bandwidth on our end

@gante Oh sorry, I was not aware. I removed flash attention but this unmasked a bug related to synchronization of my non-blocking moving of tensors (never tested without it before😁). I have added explicit synchronization right before a tensor is needed on the device or is safe to be evicted. This makes the tests pass while still keeping all the moving non-blocking. 

Some tests are still failing but they seem unrelated and/or flaky.

n17s commented:
Since the tests have moved, I have rebased to the latest.

n17s commented:
Gentle ping @ArthurZucker

ArthurZucker commented:
I was off for a week! Back now, will review!

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,110 @@ def batch_select_indices(self, indices: torch.Tensor):
+    def prefetch_layer(self, layer_idx: int):
+        "Starts prefetching the next layer cache"
+        if layer_idx < len(self):
+            with torch.cuda.stream(self.prefetch_stream):
```
does that only work when using compile or in general ?

n17s replied:
CUDA streams should work in general. torch.compile docs don't mention any known issues related to streams.

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,110 @@ def batch_select_indices(self, indices: torch.Tensor):
 
 
+class OffloadedCache(DynamicCache):
+    """
```
at some point we should warn if generating on CPU that GPU is required + if we use cuda stream this requires a hardware that supports cuda streams

n17s replied:
Hmm, the point of this class is you want to generate on the GPU but you have more CPU RAM than GPU RAM. If you are generating on the CPU you don't need this class at all.

n17s replied:
Added a check to protect this.

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -1794,6 +1795,8 @@ def generate(
                     )
 
                 model_kwargs["past_key_values"] = cache_class(cache_config)
+            elif generation_config.cache_implementation == "offloaded":
+                model_kwargs["past_key_values"] = OffloadedCache()
```
why don't we update the mapping that gives us the cahce class here?

n17s replied:
There is no such mapping? The only mappings I see in the code are
```python
 NEED_SETUP_CACHE_CLASSES_MAPPING = {"static": StaticCache, "sliding_window": SlidingWindowCache, "hybrid": HybridCache}
 QUANT_BACKEND_CLASSES_MAPPING = {"quanto": QuantoQuantizedCache, "HQQ": HQQQuantizedCache}
```
OffloadedCache does not need setup and is not a quantized cache.

ArthurZucker replied:
Yeah I was thinking about the `NEED_SETUP_CACHE_CLASSES_MAPPING`, but alrgith

ArthurZucker reviewed (commented):
Looks great, small Q about using cuda streams, we should also add this to the doc / show and example as I guess this works with compile no?

n17s commented:
> Looks great, small Q about using cuda streams, we should also add this to the doc / show and example as I guess this works with compile no?

Not sure. How to test it?

ArthurZucker reviewed (commented):
You can just compile the model's forward and do this: 
```
import os
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
os.environ["TOKENIZERS_PARALLELISM"] = "false" # silence warnings when compiling

device = "cuda"
ckpt = "model-to-test"

model = AutoModelForCausalLM.from_pretrained(ckpt, torch_dtype=torch.float16)
model.to(device)

tokenizer = AutoTokenizer.from_pretrained(ckpt)

prompt = "Why dogs are so cute?"
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")

# Specify the max length (including both the prompt and the response)
# When calling `generate` with `cache_implementation="static" later, this is also used to create a `StaticCache` object
# with sequence length = `max_length`. The longer the more you will re-use it
model.generation_config.max_length = 128

# without `torch.compile`: each call takes ~ 5.0 seconds (on A100 80G + torch 2.3)
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
print(response)

# `torch.compile(model, ...)` is not recommended as you compile callbacks
# and full generate. We recommend compiling only the forward for now. 
# "reduce-overhead" will use cudagraphs. 
model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
model.generation_config.cache_implementation = "static"

# with `torch.compile` (on A100 80G + torch 2.3)
# 1st call: ~ 90 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
# 2nd call: ~ 60 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
# 3nd call: ~ 1.5 seconds
outputs = model.generate(**inputs, do_sample=False)
response = tokenizer.batch_decode(outputs)[0]
print(response)
```
What is missing here for me is mostly raise error to protect usage (on GPU) mostly, resolve merge conflicts and should be good to go

n17s commented:
> You can just compile the model's forward and do this:

So this does not work with offloaded cache. The error message is 
```
AttributeError: 'NoneType' object has no attribute 'synchronize'
```
Stemming from
```python
    self.prefetch_stream.synchronize()
```

Based on my understanding of the discussion on [this pytorch issue](https://github.com/pytorch/pytorch/issues/92804) what is happening is that CUDA streams are not preserved by the current implementation of torch.compile. Since this seems to be a limitation with how torch.compile currently handles things, and may eventually be fixed according to the discussion, I'd say there's not much that can be done on our side.

n17s committed:
Prevent usage in CPU only mode

n17s commented:
> What is missing here for me is mostly raise error to protect usage (on GPU) mostly, resolve merge conflicts and should be good to go

Merge conflicts have been resolved and I added a check in the constructor to protect usage. Fun fact, the class works even if inference happens on the CPU, as long as there's a GPU on the machine.

ArthurZucker reviewed (commented):
Very nice! One thing missing is a bit of doc / doc with an exampe of how to easily activate this : ex with high beam search ! either in https://huggingface.co/docs/transformers/v4.43.2/en/generation_strategies or in https://huggingface.co/docs/transformers/en/perf_infer_gpu_one !

ArthurZucker commented:
(It's agreen on my side to merge otherwise!)

n17s committed:
Add a section about offloaded KV cache to the docs

n17s commented:
> One thing missing is a bit of doc / doc with an exampe of how to easily activate this
Added some docs

n17s committed:
Fix typos in docs

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
+    Useful for generating from models with very long context.
+
+    When layer k is executing it moves the cache of layer k-1 to the CPU and prefetches the KV cache of layer k+1.
+    """
```
would make sense to explain how the streams work, basically one stream for prefetching, and one stream that computes

n17s replied:
Done

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
+
+    def evict_previous_layer(self, layer_idx: int):
+        "Moves the previous layer cache to the CPU"
+        if len(self) > 2:
+            prev_layer_idx = (layer_idx - 1) % len(self)
+            self.key_cache[prev_layer_idx] = self.key_cache[prev_layer_idx].to("cpu", non_blocking=True)
+            self.value_cache[prev_layer_idx] = self.value_cache[prev_layer_idx].to("cpu", non_blocking=True)
```
can't the prefetch stream be used for that as well?

n17s replied:
It should not! Eviction needs to be queued after all other operations on these tensors happening on the default stream. If we call it from prefetch stream, it is possible that the prefetch stream moves it out (and the memory allocator puts something else in that memory) before all read operations from the default stream are done. I will add a small comment in the code about this.

ArthurZucker reviewed (commented):
Very nice!

n17s committed:
Clarifications and better explanation of streams

n17s commented on src/transformers/cache_utils.py:
```diff
@@ -447,6 +447,111 @@ def batch_select_indices(self, indices: torch.Tensor):
+    """
+
+    def __init__(self) -> None:
+        assert torch.cuda.is_available(), "OffloadedCache can only be used with a GPU"
```
Fixed

ArthurZucker reviewed (approved):
Nice thanks for iterating!

ghadiaravi13 commented:
this cache is very helpful (for low-memory long-context inference)
thanks for adding it!
